### PR TITLE
`Communication`: Enhance messages and threads user interface

### PR DESCRIPTION
**Pull request by egekurt123**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

The primary motivation behind these updates is to enhance the overall user experience within the messages/communication module. By implementing these changes, we aim to achieve better usability, modernized look and feel

### Steps for Testing

Prerequisites:
- 1 Student / Instructor / Admin
- 1 Course with Messages and Communications enabled

1. Log in to Artemis
2. Navigate to Messages
3. Make sure everything work as expected
4. Navigate to Communications
5. Make sure everything work as expected

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| course-management-tab-bar.component.ts | 93.97% | ✅ ❌ |
| course.model.ts | 100% | ✅ |
| exam-update.component.ts | 88.55% | ✅ ❌ |
| exercise-title-channel-name.component.ts | 100% | ✅ |
| lecture-title-channel-name.component.ts | 100% | ✅ |
| course-conversations.component.ts | 95.12% | ✅ ❌ |
| conversation-header.component.ts | 96.36% | ✅ ❌ |
| conversation-messages.component.ts | 94.11% | ✅ ❌ |
| metis-conversation.service.ts | 84.97% | ✅ ❌ |
| metis.service.ts | 88.47% | ✅ ❌ |
| post.component.ts | 94.91% | ✅ ❌ |
| post-footer.component.ts | 100% | ✅ |
| answer-post-header.component.ts | 96.15% | ✅ ❌ |
| post-header.component.ts | 95.65% | ✅ ❌ |
| posting-header.directive.ts | 100% | ✅ |
| posting-markdown-editor.component.ts | 97.56% | ✅ ❌ |
| post-reactions-bar.component.ts | 98.33% | ✅ ❌ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| CourseRepository.java | 76% | ✅ ❌ |
| ChannelResource.java | 91% | ✅ ❌ |
| ConversationManagementResource.java | 100% | ✅ |
| ConversationResource.java | 92% | ✅ ❌ |

�

#### Light Mode

Old 
Screenshot 2024-05-31 at 14 30 41
Screenshot 2024-05-31 at 14 29 31

New
Screenshot 2024-07-08 at 16 09 55
Screenshot 2024-07-08 at 16 14 39

#### Dark Mode

Old
Screenshot 2024-07-08 at 15 09 39
Screenshot 2024-07-08 at 15 08 34

New
Screenshot 2024-07-08 at 16 10 01
Screenshot 2024-07-08 at 16 14 46

### Notes:
- Option buttons like edit or delete message, as well as reactions will be updated in a seperate follow-up PR 

## Summary by CodeRabbit

- **New Features**
  - Enhanced post management with sorting and tagging functionalities.
  - Improved layout and conditional rendering in post and post footer components.
  - Added dynamic display for answer posts based on conditions.

- **Bug Fixes**
  - Corrected the display issues in various post components.
  
- **Style**
  - Updated badge styles with role-specific colors for students, tutors, and instructors.

- **Refactor**
  - Renamed function for clarity in communication settings within the exam update component.

### CODE REVIEW

**Review by az108 — DISMISSED** [review]
Tested on TS4 as Instructor and Student. Works as expected and I like the changes a lot 👍 . Code LGTM aswell!

**Review by Jan-Thurner — APPROVED** [review]
Tested on TS1 as an instructor and as a Student. Though the functionality works as expected, and the UI looks loads better than before, I have a few remarks:

The Distance between the two buttons on the top right feels a bit big:
image
If this view is opened on a small screen in combination with opening in thread answers, the right button disappears behind the threads:
image

The location of the `Send` button at the bottom right feels a bit odd to me too. The gap between the button and the textbox is rather big and the button is touching the right border of the window.
image

As the UI improvements are great and working, I think the suggestions can be moved into a followup PR.

**Review by sarpsahinalp — APPROVED** [review]
code lgtm

**Review by edkaya — APPROVED** [review]
code lgtm

**Comment by egekurt123** [comment]
> Tested on TS1 as an instructor and as a Student. Though the functionality works as expected, and the UI looks loads better than before, I have a few remarks:
> 
> The Distance between the two buttons on the top right feels a bit big: image If this view is opened on a small screen in combination with opening in thread answers, the right button disappears behind the threads: image
> 
> The location of the `Send` button at the bottom right feels a bit odd to me too. The gap between the button and the textbox is rather big and the button is touching the right border of the window. image
> 
> As the UI improvements are great and working, I think the suggestions can be moved into a followup PR.

Thank you for addressing those issues! I've added margin to the sides of the threads to prevent the button from sticking to the right side. Additionally, when the window is very small, the button now aligns under the channel name instead of disappearing.
Screenshot 2024-06-03 at 14 36 47

**Review by edkaya — DISMISSED** [review]
Reapprove

**Review by az108 — APPROVED** [review]
reapprove

**Review by Jan-Thurner — DISMISSED** [review]
Thanks for implementing the feedback. Tested again on TS3. Works as expected.

**Review by sarpsahinalp — DISMISSED** [review]
Reapprove

**Review by DominikRemo — CHANGES_REQUESTED** [review]
Tested on ts4
I've noticed that opening a conversation always results in the tag 'new' being displayed for other users' messages, even if the message is not new and has already been viewed.

Please also have another look at the client tests.
`AnswerPostHeaderComponent › should set author information correctly` currently fails

File: src/main/webapp/app/shared/metis/post/post.component.ts
Comment (beyzaaltuntas) [inline src/main/webapp/app/shared/metis/post/post.component.ts:133]: Shouldn't we avoid in-place sorting?
 Maybe we can implement it like that:
` this.sortedAnswerPosts = this.posting.answers.slice().sort((a, b) => 
`

File: src/main/webapp/app/shared/metis/posting-footer/post-footer/post-footer.component.scss
Comment (beyzaaltuntas) [inline src/main/webapp/app/shared/metis/posting-footer/post-footer/post-footer.component.scss:4]: Don't we have shared font sizes somewhere?

File: src/main/webapp/app/shared/metis/post/post.component.ts
Comment (beyzaaltuntas) [inline src/main/webapp/app/shared/metis/post/post.component.ts:148]: You can shorten this to `this.tags = this.posting.tags ?? []`

**Comment by egekurt123** [comment]
> Tested on ts4 I've noticed that opening a conversation always results in the tag 'new' being displayed for other users' messages, even if the message is not new and has already been viewed.
> 
> Please also have another look at the client tests. `AnswerPostHeaderComponent › should set author information correctly` currently fails

Thanks for your comment and for pointing out the bug. Although this bug was not related to this PR, it has now been fixed. The test passes successfully as well.

**Review by DominikRemo — CHANGES_REQUESTED** [review]
Thank you for fixing the client tests!
Tested on ts4.

The "New" badge does not appear at all on the communication page (intended?) and the problem persists in the messages page when switching between threads (the badges disappeared after reloading the page).

https://github.com/ls1intum/Artemis/assets/47261058/d9b0691b-bc74-4f32-b689-68874336466f

I have also noticed that the replies to a message are cut off in conversations. 
In the picture below, for example, there are only 3 replies, although there should be 9. 

Bildschirmfoto 2024-06-16 um 21 53 07

If there are a sufficient number of other messages, the responses cannot be expanded at all.

https://github.com/ls1intum/Artemis/assets/47261058/3c48357d-d381-4ff3-8871-5eebc6a60dc3

**Comment by egekurt123** [comment]
@DominikRemo Thanks for your reports.

- The "New" Badge only exists in the Messages Page
- I will take a look at the case without refreshing the page
- The UI Problem that causes messages to get cut off is a consequence of the deprecated paging in the communications page; my UI changes just apparently make it a lot worse. This is why I decided to merge this PR after #8801 which merges the Messages and Communication modules

**Review by az108 — DISMISSED** [review]
reapprove after merge conflict

**Review by edkaya — APPROVED** [review]
reapprove after merge conflict

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]
The roles are assigned more or less randomly in the thread ui

**Review by SimonEntholzer — CHANGES_REQUESTED** [review]
Tested during testing session, new changes look nice, only issue we found was that student/tutor tags over messages are not always shown correctly.

**Review by JohannesStoehr — APPROVED** [review]
Code with fix looks good to me now

**Review by Jan-Thurner — CHANGES_REQUESTED** [review]
When writing messages as an intsructor, only the newest one is marked as an instructor. The other messages are marked as a student. When reloading the page, all messages are marked as instructor again.
image
After reload:
image

File: src/main/webapp/app/shared/metis/posting-header/post-header/post-header.component.html
Comment (rabeatwork) [inline src/main/webapp/app/shared/metis/posting-header/post-header/post-header.component.html:3]: See Feedback in src/main/webapp/app/shared/metis/posting-header/answer-post-header/answer-post-header.component.html

File: src/main/webapp/app/shared/metis/posting-header/post-header/post-header.component.html
Comment (rabeatwork) [inline src/main/webapp/app/shared/metis/posting-header/post-header/post-header.component.html:1]: Not part of this PR, but it would make sense to create a component for it. Since we use more or less the same code in several files.

File: src/main/webapp/content/scss/global.scss
Comment (rabeatwork) [inline src/main/webapp/content/scss/global.scss:1095]: Great idea to add global classes for these font sizes. However, please use either rem values or the corresponding keywords like `small` or `smaller` and so on. rem is relative to the font size of the root element. In our case 1 rem corresponds to 16px. The keywords correspond to the following rem values:

xx-small: 0.625rem
x-small: 0.75rem
small: 0.875rem
medium: 1rem
large: 1.125rem
x-large: 1.5rem
xx-large: 2rem
xxx-large: 3rem

File: src/main/webapp/app/shared/metis/posting-header/post-header/post-header.component.html
Comment (rabeatwork) [inline src/main/webapp/app/shared/metis/posting-header/post-header/post-header.component.html:32]: Please make sure that these icons don't wrap 
image

**Review by rabeatwork — CHANGES_REQUESTED** [review]

File: src/main/webapp/app/overview/discussion-section/discussion-section.component.scss
Comment (egekurt123) [reply src/main/webapp/app/overview/discussion-section/discussion-section.component.scss]: I changed it again, but the new font sizes and badges are causing an extra line break in the message header when users have longer names. We can look at it again and add a quick followup PR

File: src/main/webapp/app/overview/course-conversations/layout/conversation-thread-sidebar/conversation-thread-sidebar.component.scss
Comment (egekurt123) [reply src/main/webapp/app/overview/course-conversations/layout/conversation-thread-sidebar/conversation-thread-sidebar.component.scss]: I changed it again, but the new font sizes and badges are causing an extra line break in the message header when users have longer names. We can look at it again and add a quick followup PR

**Review by pzdr7 — COMMENTED** [review]
Tested on TS5. 

I'm encountering an issue where the edit and delete icons don't match the messages in a thread.

To reproduce:

User 1 (instructor): Send two messages in a channel. On the first message, reply a few times.
User 2 (student): On the second message, reply a few times. Then, switch to the thread of the first message.

The delete/edit icons stay. In this example, I can see them as test user 3 even though the messages were written by test user 16.

image

(I assume it doesn't show up for the bottom one because the other thread, where I replied with test user 3, had just two replies instead of three).

https://artemis-test5.artemis.cit.tum.de/courses/20/communication?conversationId=226

**Comment by egekurt123** [comment]
> Tested on TS5.
> 
> I'm encountering an issue where the edit and delete icons don't match the messages in a thread.
> 
> To reproduce:
> 
> User 1 (instructor): Send two messages in a channel. On the first message, reply a few times. User 2 (student): On the second message, reply a few times. Then, switch to the thread of the first message.
> 
> The delete/edit icons stay. In this example, I can see them as test user 3 even though the messages were written by test user 16.
> 
> image
> 
> (I assume it doesn't show up for the bottom one because the other thread, where I replied with test user 3, had just two replies instead of three).
> 
> https://artemis-test5.artemis.cit.tum.de/courses/20/communication?conversationId=226

I have just took a deeper look into this and found that it can also be reproduced on other test servers, confirming that it is an existing bug.

**Review by JohannesStoehr — APPROVED** [review]
Code

**Review by pzdr7 — APPROVED** [review]

**Review by Jan-Thurner — APPROVED** [review]
Thanks for looking into the issue and fixing it. LGTM now.

**Review by az108 — APPROVED** [review]
Changes LGTM

**Review by bassner — CHANGES_REQUESTED** [review]
Please add the test coverage table to the PR

**Comment by bassner** [comment]
@egekurt123 You did not remove the hardcoded colors 🫣 please read the docs i linked in my comment

**Comment by bassner** [comment]
@egekurt123 You did not remove the hardcoded colors 🫣 please read the docs i linked in my comment

**Pull request merged**
